File: c03_3_binary_search.py
import unittest

def method1(array,x):
    a,b=0,len(array)-1
    while a<=b:
        k=(a+b)//2
        if array[k]==x:
            return k
        if array[k]>x:
            b=k-1
        else:
            a=k+1
    return -1

def method2(array,x):
    n=len(array)
    k,b=0,n//2
    while b>=1:
        while k+b<n and array[k+b]<=x:
            k+=b
        b//=2
    return k if n>0 and array[k]==x else -1

def smallest(z,ok):
    x,b=-1,z
    while b>=1:
        while not ok(x+b):
            x+=b
        b//=2
    return x+1

def maximum(z,f):
    x,b=-1,z
    while b>=1:
        while f(x+b)<f(x+b+1):
            x+=b
        b//=2
    return x+1

class Test(unittest.TestCase):
    def test(self):
        for m in (method1,method2):
            with self.subTest(m=m):
                self.assertEqual(m([],5),-1)
                self.assertEqual(m([*range(10)],5),5)
                self.assertEqual(m([*range(10)],10),-1)
        self.assertEqual(smallest(10,lambda x:x>4),5)
        self.assertEqual(maximum(10,lambda x:-abs(5-x)),5)

if __name__=='__main__':
    unittest.main()

File: c++/c03_3_binary_search.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	ll method1(const vector<ll>& array, ll x) {
		ll n = ssize(array);
		ll a = 0, b = n - 1;
		while (a <= b) {
			ll k = (a + b) / 2;
			if (array[k] == x) {
				// x found at index k
				return k;
			}
			if (array[k] > x) b = k - 1;
			else a = k + 1;
		}
		return -1;
	}

	ll method2(const vector<ll>& array, ll x) {
		ll n = ssize(array);
		ll k = 0;
		for (ll b = n / 2; b >= 1; b /= 2) {
			while (k + b < n && array[k + b] <= x) k += b;
		}
		if (k < ssize(array) && array[k] == x) {
			// x found at index k
			return k;
		}
		return -1;
	}

	ll smallest(ll z, auto ok) {
		ll x = -1;
		for (ll b = z; b >= 1; b /= 2) {
			while (!ok(x + b)) x += b;
		}
		return x + 1;
	}

	ll maximum(ll z, auto f) {
		ll x = -1;
		for (ll b = z; b >= 1; b /= 2) {
			while (f(x + b) < f(x + b + 1)) x += b;
		}
		return x + 1;
	}

	void test(auto f) {
		EXPECT_EQ(f({}, 5), -1);
		EXPECT_EQ(f({ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }, 5), 5);
		EXPECT_EQ(f({ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }, 10), -1);
	}
}

TEST(C033BinarySearch, method1) {
	test(method1);
}

TEST(C033BinarySearch, method2) {
	test(method2);
}

TEST(C033BinarySearch, smallest) {
	EXPECT_EQ(smallest(10, [](ll x) { return x > 4; }), 5);
}

TEST(C033BinarySearch, maximum) {
	EXPECT_EQ(maximum(10, [](ll x) { return -abs(5 - x); }), 5);
}
